---
title: "CRUK CI Summer School 2020 - introduction to single-cell RNA-seq analysis"
subtitle: 'cellranger'

author: "Stephane Ballereau & Ashley Sawle"
output:
  html_document:
    df_print: paged
    toc: yes
    number_sections: true
    code_folding: hide
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE, cache=FALSE)
knitr::opts_chunk$set(echo=FALSE)
```

# Alignment and feature counting {#AliFeatCountTop}

## Introduction

The first step in the analysis of single cell RNAseq data is to align the
sequences reads against a genomic reference and then use a transcriptome
annotation to generate read counts for each features of interest. Typically
for scRNAseq the features of interest are genes.

There are a variety of tools for doing this and your choice will depend in part
on the method by which the library was generated. For data generated using the
10x-Chromium method data the most common approach is to use the 
[Cell Ranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger)
tool provided by 10x. This not only carries out the alignment and feature
counting, but will also:

* Call cells, i.e. filter the raw matrix to remove droplets that do not contain
  cells   
* Generate a very useful report in html format, which will provide some QC
  metrics and an initial look at the data   
* Generate a "cloupe" file, which can be opened using the [10x loupe
  browser](https://support.10xgenomics.com/single-cell-gene-expression/software/visualization/latest/what-is-loupe-cell-browser)
  software to further explore the data.   

Alternative methods include:

* **[STAR solo](https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md)** -
  this tool is built into the general purpose STAR aligner (Cell Ranger actually
  uses STAR under the hood). This will generate outputs very similar to
  CellRanger minus the cloupe file and the QC report. The advantage over
  CellRanger is that it is much more "lightweight" and will run with lower
  memory requirements in a shorter time.
* **[Alevin](https://salmon.readthedocs.io/en/latest/alevin.html)** - This tool
  is based on the popular Salmon tool for bulk RNAseq feature counting..............
  Alevin supports both 10x-Chromium and Drop-seq derived data ..................   
* Something about DropSeq etc. ................   
* Any other....................???     

For the purposes of this course, seeing as we are working with 10x-Chromium
derived data, we will use Cell Ranger. As detailed instructions are available
on the Cell Ranger pages of the 10x website, this chapter will not be
comprehensive in terms of all option, but should provide a bried overview.

## 10X Cell Ranger pipeline in brief

The Cell Ranger tools actually incorporates a number of different analysis
pipelines for handling different components of the single cell RNAseq analysis.
In this chapter we will be using the `count` tool, later in the course you will
encounter the `aggr` (aggregate) tool, which can be used to merge mutliple
samples into a single experiment-level data set. 

In addition to the analysis tools, Cell Ranger also includes a tools to
generate a custom Cell Ranger reference from genomic and transcriptomic
references.

### Installing Cell Ranger

Full installation instruction can be found on the 
[10x website](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/installation).

### Cell Ranger references

The Cell Ranger software, like other aligners, requires the information about
the genome and transcriptome of interest to be provided in a specific format.
If you are working with the standard genome/transcriptome of human or mouse
then you can download prebuilt references from the [10x
website](https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest)
(there is even a combined genome reference, which is useful for PDX based
experiments).

If you require a reference for the genome of another species, or wish to add 
custom transcripts to the gene annotation (e.g. fusion genes), then you will 
need to build your own index. We will not go into detail here about this, but
it is a relatively straightforward process and you can find instructions 
[here](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/references).

### Cell Ranger count

Full details can be found on the 
[10x website](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_ct).

#### Preparing the raw fastq files

To run Cell Ranger the fastq files for samples to be processed should be placed
in a single directory. Cell Ranger will be run separately on each file, you
will need to provide the software with the index (e.g. SIGAA6) of the sample to
be processed. Cell Ranger expects the file names of the fastq files to follow a
specific convention so that it can correctly identify the files for the
specific sample. If the files names of your fastqs do not match this convention
you will need to modify them.

The convention is the default convention used by `bcl2fastq` which processes
the raw sequencing data into fastq format:

```
<SampleName>_S<SampleNumber>_L00<Lane>_<Read>_001.fastq.gz
```

* **\<SampleName\>** - An identifier for the sample, this is what Cell Ranger
  uses to determine which fastq files to combine into a single sample.
* **\<SampleNumber\>** - This is the sample number based on the order that
  samples were listed in the sample sheet used when running `bcl2fastq`. This
  is not important for Cell Ranger, other than it indicates the end of the
  Sample Name, you can set all your samples to S1.
* **\<Lane\>** - The lane number. If your sequencing was run across mutliple
  lanes, then you may have multiple sets of fastqs for a single sample with
  different lane numbers.
* **\<Read\>** - The read. **R1** for Read 1, **R2** for Read 2, and index
  reads are **I1** and **I2**.
* **001**  - The last segment is always 001.

e.g. for a single samples in the Caron data set we have:

```
    SRR9264343_S0_L001_I1_001.fastq.gz
    SRR9264343_S0_L001_R1_001.fastq.gz
    SRR9264343_S0_L001_R2_001.fastq.gz
```

#### Running `cellranger count`

The minimum information require to run `cellranger count` is:

* **--id** - A sample ID. This is used for naming the outputs   
* **--transcriptome** - the directory containing the Cell Ranger reference   
* **--fastqs** - the directory containing the fastq files

This will process all fastq files in the `fastqs` directory into a single
sample. If you have multiple samples in a single directory then you need to
add:

* **--sample** - the SampleName from the fastq files as above.

In addition, Cell Ranger is very computationally intensive, you will usually
be wanting to run it on a high performance cluster or server, and it is 
advisable to set limits to the resources it will use:

* **--localcores** - the number of processors Cell Ranger should use   
* **--localmem** - the amount of memory, in Gigabytes, Cell Ranger should use.

e.g:

```
cellranger count --id=SRR9264343 \
                 --transcriptome=refdata-gex-mm10-2020-A \
                 --fastqs=fastq \
                 --sample=SRR9264343 \
                 --localcores=8 \
                 --localmem=64
```

### Cell Ranger outputs

Cell Ranger will create a single folder for the sample names as per the `--id` option
in the command. In the top level of the directory is a directory called `outs`, this 
contains the results of the analysis pipeline. There are also many intermediate
and log files/directories that will not be of immediate interest.

The contents of the sample directory will look like this:

![](Images/CellRangerOutput.png){width=50%}

The contents of the output directory are:

![](Images/CellRangerOutputOuts.png){width=50%}

* **analysis** - The results of clustering and differential expression analysis
  on the clusters. These are used in the *web_summary.html* report.
* **cloupe.cloupe** - a cloupe file for loading into the 10x loupe browser
* **filtered_feature_bc_matrix** - The filtered count matrix directory 
* **filtered_feature_bc_matrix.h5** - The filtered count matrix as an HDF5 file
* **metrics_summary.csv** - summary metrics from the analysis  
* **molecule_info.h5** - per-molecule read information as an HDF5 file 
* **possorted_genome_bam.bam** - The aligned reads in bam format 
* **possorted_genome_bam.bam.bai** - The bam index 
* **raw_feature_bc_matrix** - The raw count matrix directory 
* **raw_feature_bc_matrix.h5** - The raw count matrix as an HDF5 file 
* **web_summary.html** - The summary report 

The two count matrix directories each contain 3 files:

* **barcodes.tsv.gz** - The cell barcodes detected; these correspond to the
  columns of the count matrix
* **features.tsv.gz** - The features detected. In this cases gene ids. These
  correspond to the rows of the count matrix.
* **matrix.mtx.gz** - the count of unique UMIs for each gene in each cell.

The filtered count matrix only contains reads that have been called as cells
by CellRanger.

#### What are the counts?

Remember that for each read we have a sample barcode, a cell barcode, and a
UMI.  In addition we now have location on the genome and a corresponding gene
annotation.  The count for any gene for any cell is the number of unique reads
detected for the combination:

sample barcode + cell barcode + UMI + gene ID

Thus we avoid counting PCR duplicate reads as these will have the same UMI. For
this reason we more commmonly talk about the "UMI count" rather than the "Read
count"; the "Read count" could be higher than the "UMI count"




